### ISSUE TITLE
virtio-fs: Unable to copy large file

### ISSUE THREAD

**Issue by kevinoid** [virtio-win/kvm-guest-drivers-windows#527]
When sharing a host directory with a Windows 10 (10.0.19042 Build 19042) guest via [virtio-fs](https://virtio-fs.gitlab.io) using `viofs.sys` and `virtiofs.exe` from [`virtio-win-0.1.190.iso`](https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/archive-virtio/virtio-win-0.1.190-1/virtio-win-0.1.190.iso) and [WinFSP 2020.2](https://github.com/billziss-gh/winfsp/releases/tag/v1.8), when I attempt to copy a file which is more than 2,097,152 bytes, the copy fails.  For example, on the host:
```sh
dd if=/dev/zero of=bigfile.bin bs=2097153 count=1
```

On the guest:

    > copy Z:\bigfile.bin .
    A device attached to the system is not functioning.
            0 file(s) copied.

<details>
<summary><code>virtiofs.exe -d -1</code> debug log</summary>

    *** VirtFsFuseRequest: >>req: 26 unique: 2 len: 56
    *** VirtFsFuseRequest: <<len: 80 error: 0 unique: 2
    virtiofs[TID=1080]: FFFFCE8A0FC4D010: >>QueryVolumeInformation
    *** VirtFsFuseRequest: >>req: 17 unique: 3 len: 40
    *** VirtFsFuseRequest: <<len: 96 error: 0 unique: 3
    *** GetVolumeInfo: VolumeLabel: Kevin
    virtiofs[TID=1080]: FFFFCE8A0FC4D010: <<QueryVolumeInformation IoStatus=0[0] VolumeInfo={TotalSize=71:aecea000, FreeSize=19:8382000, VolumeLabel="Kevin"}
    virtiofs[TID=1080]: FFFFCE8A1038A010: >>Create [UT---C] "\autorun.inf", FILE_OPEN, CreateOptions=200000, FileAttributes=0, Security=NULL, AllocationSize=0:0, AccessToken=00000000000001D4[PID=1230], DesiredAccess=80, GrantedAccess=0, ShareAccess=7
    *** GetSecurityByName: "\autorun.inf"
    *** VirtFsFuseRequest: >>req: 1 unique: 4 len: 52
    *** VirtFsFuseRequest: <<len: 16 error: -2 unique: 4
    virtiofs[TID=1080]: FFFFCE8A1038A010: <<Create IoStatus=c0000034[0]
    virtiofs[TID=1080]: FFFFCE8A1038A010: >>Create [UT---C] "\autorun.inf", FILE_OPEN, CreateOptions=60, FileAttributes=0, Security=NULL, AllocationSize=0:0, AccessToken=00000000000001D4[PID=1230], DesiredAccess=120089, GrantedAccess=0, ShareAccess=7
    *** GetSecurityByName: "\autorun.inf"
    *** VirtFsFuseRequest: >>req: 1 unique: 5 len: 52
    *** VirtFsFuseRequest: <<len: 16 error: -2 unique: 5
    virtiofs[TID=1080]: FFFFCE8A1038A010: <<Create IoStatus=c0000034[0]
    virtiofs[TID=1080]: FFFFCE8A1038A010: >>Create [UT---C] "\autorun.inf", FILE_OPEN, CreateOptions=60, FileAttributes=0, Security=NULL, AllocationSize=0:0, AccessToken=00000000000001D4[PID=1230], DesiredAccess=120089, GrantedAccess=0, ShareAccess=7
    *** GetSecurityByName: "\autorun.inf"
    *** VirtFsFuseRequest: >>req: 1 unique: 6 len: 52
    *** VirtFsFuseRequest: <<len: 16 error: -2 unique: 6
    virtiofs[TID=1080]: FFFFCE8A1038A010: <<Create IoStatus=c0000034[0]
    virtiofs[TID=1080]: FFFFCE8A110DD940: >>Create [UT---C] "\", FILE_OPEN, CreateOptions=200000, FileAttributes=0, Security=NULL, AllocationSize=0:0, AccessToken=00000000000001D4[PID=1230], DesiredAccess=80, GrantedAccess=0, ShareAccess=7
    *** GetSecurityByName: "\"
    *** VirtFsFuseRequest: >>req: 1 unique: 7 len: 42
    *** VirtFsFuseRequest: <<len: 144 error: 0 unique: 7
    *** SubmitLookupRequest: nodeid=4 ino=1859626 size=4096 blocks=8 atime=1609268420 mtime=1609268413 ctime=1609268413 atimensec=502884297 mtimensec=38760563 ctimensec=38760563 mode=41ed nlink=2 uid=1000 gid=1000 rdev=0 blksize=4096
    *** Open: "\" CreateOptions: 0x01200000 GrantedAccess: 0x00000080
    *** VirtFsFuseRequest: >>req: 1 unique: 8 len: 42
    *** VirtFsFuseRequest: <<len: 144 error: 0 unique: 8
    *** SubmitLookupRequest: nodeid=4 ino=1859626 size=4096 blocks=8 atime=1609268420 mtime=1609268413 ctime=1609268413 atimensec=502884297 mtimensec=38760563 ctimensec=38760563 mode=41ed nlink=2 uid=1000 gid=1000 rdev=0 blksize=4096
    *** VirtFsFuseRequest: >>req: 27 unique: 9 len: 48
    *** VirtFsFuseRequest: <<len: 32 error: 0 unique: 9
    *** Open: fh: 0 nodeid: 4
    *** SetFileInfo: ino=1859626 size=4096 blocks=8 atime=1609268420 mtime=1609268413 ctime=1609268413 atimensec=502884297 mtimensec=38760563 ctimensec=38760563 mode=41ed nlink=2 uid=1000 gid=1000 rdev=0 blksize=4096
    virtiofs[TID=1080]: FFFFCE8A110DD940: <<Create IoStatus=0[1] UserContext=0000000000000000:000002231FC0AF30, GrantedAccess=80, FileInfo={FileAttributes=10, ReparseTag=0, AllocationSize=0:1000, FileSize=0:1000, CreationTime=2020-12-29T19:00:13.038Z, LastAccessTime=2020-12-29T19:00:20.502Z, LastWriteTime=2020-12-29T19:00:13.038Z, ChangeTime=2020-12-29T19:00:13.038Z, IndexNumber=0:0}
    virtiofs[TID=1080]: FFFFCE8A100F65B0: >>Close 0000000000000000:000002231FC0AF30
    *** Close: fh: 0 nodeid: 4
    *** VirtFsFuseRequest: >>req: 29 unique: 10 len: 64
    *** VirtFsFuseRequest: <<len: 16 error: 0 unique: 10
    virtiofs[TID=1080]: FFFFCE8A100F65B0: <<Close IoStatus=0[0]
    virtiofs[TID=1080]: FFFFCE8A0FC4D010: >>Create [UT---C] "\AutoRun.inf", FILE_OPEN, CreateOptions=200000, FileAttributes=0, Security=NULL, AllocationSize=0:0, AccessToken=00000000000001D4[PID=1230], DesiredAccess=80, GrantedAccess=0, ShareAccess=7
    *** GetSecurityByName: "\AutoRun.inf"
    *** VirtFsFuseRequest: >>req: 1 unique: 11 len: 52
    *** VirtFsFuseRequest: <<len: 16 error: -2 unique: 11
    virtiofs[TID=1080]: FFFFCE8A0FC4D010: <<Create IoStatus=c0000034[0]
    virtiofs[TID=1080]: FFFFCE8A1038A010: >>Create [UT---C] "\bigfile.bin", FILE_OPEN, CreateOptions=200000, FileAttributes=0, Security=NULL, AllocationSize=0:0, AccessToken=00000000000001D4[PID=11f4], DesiredAccess=80, GrantedAccess=0, ShareAccess=7
    *** GetSecurityByName: "\bigfile.bin"
    *** VirtFsFuseRequest: >>req: 1 unique: 12 len: 52
    *** VirtFsFuseRequest: <<len: 144 error: 0 unique: 12
    *** SubmitLookupRequest: nodeid=3 ino=1836112 size=2097153 blocks=4104 atime=1609268789 mtime=1609268786 ctime=1609268786 atimensec=150215756 mtimensec=250078608 ctimensec=250078608 mode=81a4 nlink=1 uid=1000 gid=1000 rdev=0 blksize=4096
    *** Open: "\bigfile.bin" CreateOptions: 0x01200000 GrantedAccess: 0x00000080
    *** VirtFsFuseRequest: >>req: 1 unique: 13 len: 52
    *** VirtFsFuseRequest: <<len: 144 error: 0 unique: 13
    *** SubmitLookupRequest: nodeid=3 ino=1836112 size=2097153 blocks=4104 atime=1609268789 mtime=1609268786 ctime=1609268786 atimensec=150215756 mtimensec=250078608 ctimensec=250078608 mode=81a4 nlink=1 uid=1000 gid=1000 rdev=0 blksize=4096
    *** VirtFsFuseRequest: >>req: 14 unique: 14 len: 48
    *** VirtFsFuseRequest: <<len: 32 error: 0 unique: 14
    *** Open: fh: 1 nodeid: 3
    *** SetFileInfo: ino=1836112 size=2097153 blocks=4104 atime=1609268789 mtime=1609268786 ctime=1609268786 atimensec=150215756 mtimensec=250078608 ctimensec=250078608 mode=81a4 nlink=1 uid=1000 gid=1000 rdev=0 blksize=4096
    virtiofs[TID=1080]: FFFFCE8A1038A010: <<Create IoStatus=0[1] UserContext=0000000000000000:000002231FC0B110, GrantedAccess=80, FileInfo={FileAttributes=20, ReparseTag=0, AllocationSize=0:201000, FileSize=0:200001, CreationTime=2020-12-29T19:06:26.250Z, LastAccessTime=2020-12-29T19:06:29.150Z, LastWriteTime=2020-12-29T19:06:26.250Z, ChangeTime=2020-12-29T19:06:26.250Z, IndexNumber=0:0}
    virtiofs[TID=04c8]: FFFFCE8A0FF14980: >>Close 0000000000000000:000002231FC0B110
    *** Close: fh: 1 nodeid: 3
    *** VirtFsFuseRequest: >>req: 18 unique: 15 len: 64
    virtiofs[TID=1080]: FFFFCE8A1038A010: >>Create [UT---C] "\", FILE_OPEN, CreateOptions=21, FileAttributes=10, Security=NULL, AllocationSize=0:0, AccessToken=00000000000001D4[PID=11f4], DesiredAccess=100001, GrantedAccess=0, ShareAccess=7
    *** GetSecurityByName: "\"
    *** VirtFsFuseRequest: <<len: 16 error: 0 unique: 15
    *** VirtFsFuseRequest: >>req: 1 unique: 16 len: 42
    virtiofs[TID=04c8]: FFFFCE8A0FF14980: <<Close IoStatus=0[0]
    *** VirtFsFuseRequest: <<len: 144 error: 0 unique: 16
    *** SubmitLookupRequest: nodeid=4 ino=1859626 size=4096 blocks=8 atime=1609268420 mtime=1609268413 ctime=1609268413 atimensec=502884297 mtimensec=38760563 ctimensec=38760563 mode=41ed nlink=2 uid=1000 gid=1000 rdev=0 blksize=4096
    *** Open: "\" CreateOptions: 0x01000021 GrantedAccess: 0x00100001
    *** VirtFsFuseRequest: >>req: 1 unique: 17 len: 42
    *** VirtFsFuseRequest: <<len: 144 error: 0 unique: 17
    *** SubmitLookupRequest: nodeid=4 ino=1859626 size=4096 blocks=8 atime=1609268420 mtime=1609268413 ctime=1609268413 atimensec=502884297 mtimensec=38760563 ctimensec=38760563 mode=41ed nlink=2 uid=1000 gid=1000 rdev=0 blksize=4096
    *** VirtFsFuseRequest: >>req: 27 unique: 18 len: 48
    *** VirtFsFuseRequest: <<len: 32 error: 0 unique: 18
    *** Open: fh: 0 nodeid: 4
    *** SetFileInfo: ino=1859626 size=4096 blocks=8 atime=1609268420 mtime=1609268413 ctime=1609268413 atimensec=502884297 mtimensec=38760563 ctimensec=38760563 mode=41ed nlink=2 uid=1000 gid=1000 rdev=0 blksize=4096
    virtiofs[TID=1080]: FFFFCE8A1038A010: <<Create IoStatus=0[1] UserContext=0000000000000000:000002231FC0AF30, GrantedAccess=100001, FileInfo={FileAttributes=10, ReparseTag=0, AllocationSize=0:1000, FileSize=0:1000, CreationTime=2020-12-29T19:00:13.038Z, LastAccessTime=2020-12-29T19:00:20.502Z, LastWriteTime=2020-12-29T19:00:13.038Z, ChangeTime=2020-12-29T19:00:13.038Z, IndexNumber=0:0}
    virtiofs[TID=1080]: FFFFCE8A1038A010: >>QueryDirectory 0000000000000000:000002231FC0AF30, Address=000002231FE90000, Length=16368, Pattern=NULL, Marker=NULL
    *** ReadDirectory: Pattern: (null) Marker: (null) BufferLength: 16368
    *** VirtFsFuseRequest: >>req: 44 unique: 19 len: 80
    *** VirtFsFuseRequest: <<len: 680 error: 0 unique: 19
    *** VirtFsFuseRequest: Bytes Returned: 680 Expected: 32752
    *** ReadDirectory: ino=1838117 off=3420108589332170102 namelen=18 type=8 name=pcmover_pue_ml.exe
    *** ReadDirectory: "pcmover_pue_ml.exe" (18)
    *** SetFileInfo: ino=1838117 size=285388584 blocks=557408 atime=1609268420 mtime=1609268162 ctime=1609268410 atimensec=538884894 mtimensec=738640952 ctimensec=554719395 mode=81c0 nlink=1 uid=1000 gid=1000 rdev=0 blksize=4096
    *** ReadDirectory: ino=1836112 off=7210103079310993781 namelen=11 type=8 name=bigfile.bin
    *** ReadDirectory: "bigfile.bin" (11)
    *** SetFileInfo: ino=1836112 size=2097153 blocks=4104 atime=1609268789 mtime=1609268786 ctime=1609268786 atimensec=150215756 mtimensec=250078608 ctimensec=250078608 mode=81a4 nlink=1 uid=1000 gid=1000 rdev=0 blksize=4096
    *** ReadDirectory: ino=1859626 off=9204983519520793567 namelen=1 type=4 name=.
    *** ReadDirectory: "." (1)
    *** SetFileInfo: ino=1859626 size=0 blocks=0 atime=0 mtime=0 ctime=0 atimensec=0 mtimensec=0 ctimensec=0 mode=4000 nlink=0 uid=0 gid=0 rdev=0 blksize=0
    *** ReadDirectory: ino=1835009 off=9223372036854775807 namelen=2 type=4 name=..
    *** ReadDirectory: ".." (2)
    *** SetFileInfo: ino=1835009 size=0 blocks=0 atime=0 mtime=0 ctime=0 atimensec=0 mtimensec=0 ctimensec=0 mode=4000 nlink=0 uid=0 gid=0 rdev=0 blksize=0
    *** VirtFsFuseRequest: >>req: 44 unique: 20 len: 80
    *** VirtFsFuseRequest: <<len: 16 error: 0 unique: 20
    virtiofs[TID=1080]: FFFFCE8A1038A010: <<QueryDirectory IoStatus=0[498]
    virtiofs[TID=1080]: FFFFCE8A1038A010: >>Create [UT---C] "\bigfile.bin", FILE_OPEN, CreateOptions=64, FileAttributes=0, Security=NULL, AllocationSize=0:0, AccessToken=00000000000001D4[PID=11f4], DesiredAccess=120089, GrantedAccess=0, ShareAccess=3
    *** GetSecurityByName: "\bigfile.bin"
    *** VirtFsFuseRequest: >>req: 1 unique: 21 len: 52
    *** VirtFsFuseRequest: <<len: 144 error: 0 unique: 21
    *** SubmitLookupRequest: nodeid=3 ino=1836112 size=2097153 blocks=4104 atime=1609268789 mtime=1609268786 ctime=1609268786 atimensec=150215756 mtimensec=250078608 ctimensec=250078608 mode=81a4 nlink=1 uid=1000 gid=1000 rdev=0 blksize=4096
    *** Open: "\bigfile.bin" CreateOptions: 0x01000064 GrantedAccess: 0x00120089
    *** VirtFsFuseRequest: >>req: 1 unique: 22 len: 52
    *** VirtFsFuseRequest: <<len: 144 error: 0 unique: 22
    *** SubmitLookupRequest: nodeid=3 ino=1836112 size=2097153 blocks=4104 atime=1609268789 mtime=1609268786 ctime=1609268786 atimensec=150215756 mtimensec=250078608 ctimensec=250078608 mode=81a4 nlink=1 uid=1000 gid=1000 rdev=0 blksize=4096
    *** VirtFsFuseRequest: >>req: 14 unique: 23 len: 48
    *** VirtFsFuseRequest: <<len: 32 error: 0 unique: 23
    *** Open: fh: 1 nodeid: 3
    *** SetFileInfo: ino=1836112 size=2097153 blocks=4104 atime=1609268789 mtime=1609268786 ctime=1609268786 atimensec=150215756 mtimensec=250078608 ctimensec=250078608 mode=81a4 nlink=1 uid=1000 gid=1000 rdev=0 blksize=4096
    virtiofs[TID=1080]: FFFFCE8A1038A010: <<Create IoStatus=0[1] UserContext=0000000000000000:000002231FC0B110, GrantedAccess=120089, FileInfo={FileAttributes=20, ReparseTag=0, AllocationSize=0:201000, FileSize=0:200001, CreationTime=2020-12-29T19:06:26.250Z, LastAccessTime=2020-12-29T19:06:29.150Z, LastWriteTime=2020-12-29T19:06:26.250Z, ChangeTime=2020-12-29T19:06:26.250Z, IndexNumber=0:0}
    virtiofs[TID=1080]: FFFFCE8A0FF14980: >>QuerySecurity 0000000000000000:000002231FC0B110
    *** GetSecurity: fh: 1 nodeid: 3
    *** VirtFsFuseRequest: >>req: 3 unique: 24 len: 56
    *** VirtFsFuseRequest: <<len: 120 error: 0 unique: 24
    virtiofs[TID=1080]: FFFFCE8A0FF14980: <<QuerySecurity IoStatus=0[0] Security="O:S-1-5-21-3646267360-1433664085-1617066045-1001G:S-1-5-21-3646267360-1433664085-1617066045-513D:P(A;;0x1f01bf;;;S-1-5-21-3646267360-1433664085-1617066045-1001)(A;;0x1200a9;;;S-1-5-21-3646267360-1433664085-1617066045-513)(A;;0x1200a9;;;WD)"
    virtiofs[TID=1080]: FFFFCE8A1038A010: >>Read 0000000000000000:000002231FC0B110, Address=000002231FE90000, Offset=0:0, Length=512, Key=0
    *** Read: Offset: 0 Length: 512
    *** Read: fh: 1 nodeid: 3
    *** VirtFsFuseRequest: >>req: 15 unique: 25 len: 552
    *** VirtFsFuseRequest: <<len: 528 error: 0 unique: 25
    *** Read: BytesTransferred: 512
    virtiofs[TID=1080]: FFFFCE8A1038A010: <<Read IoStatus=0[512]
    virtiofs[TID=1080]: FFFFCE8A1038A010: >>Create [UT---C] "\bigfile.bin", FILE_OPEN, CreateOptions=200064, FileAttributes=0, Security=NULL, AllocationSize=0:0, AccessToken=00000000000001D4[PID=11f4], DesiredAccess=120089, GrantedAccess=0, ShareAccess=5
    *** GetSecurityByName: "\bigfile.bin"
    *** VirtFsFuseRequest: >>req: 1 unique: 26 len: 52
    *** VirtFsFuseRequest: <<len: 144 error: 0 unique: 26
    *** SubmitLookupRequest: nodeid=3 ino=1836112 size=2097153 blocks=4104 atime=1609268789 mtime=1609268786 ctime=1609268786 atimensec=150215756 mtimensec=250078608 ctimensec=250078608 mode=81a4 nlink=1 uid=1000 gid=1000 rdev=0 blksize=4096
    *** Open: "\bigfile.bin" CreateOptions: 0x01200064 GrantedAccess: 0x00120089
    *** VirtFsFuseRequest: >>req: 1 unique: 27 len: 52
    *** VirtFsFuseRequest: <<len: 144 error: 0 unique: 27
    *** SubmitLookupRequest: nodeid=3 ino=1836112 size=2097153 blocks=4104 atime=1609268789 mtime=1609268786 ctime=1609268786 atimensec=150215756 mtimensec=250078608 ctimensec=250078608 mode=81a4 nlink=1 uid=1000 gid=1000 rdev=0 blksize=4096
    *** VirtFsFuseRequest: >>req: 14 unique: 28 len: 48
    *** VirtFsFuseRequest: <<len: 32 error: 0 unique: 28
    *** Open: fh: 0 nodeid: 3
    *** SetFileInfo: ino=1836112 size=2097153 blocks=4104 atime=1609268789 mtime=1609268786 ctime=1609268786 atimensec=150215756 mtimensec=250078608 ctimensec=250078608 mode=81a4 nlink=1 uid=1000 gid=1000 rdev=0 blksize=4096
    virtiofs[TID=1080]: FFFFCE8A1038A010: <<Create IoStatus=0[1] UserContext=0000000000000000:000002231FC0B1A0, GrantedAccess=120089, FileInfo={FileAttributes=20, ReparseTag=0, AllocationSize=0:201000, FileSize=0:200001, CreationTime=2020-12-29T19:06:26.250Z, LastAccessTime=2020-12-29T19:06:29.150Z, LastWriteTime=2020-12-29T19:06:26.250Z, ChangeTime=2020-12-29T19:06:26.250Z, IndexNumber=0:0}
    virtiofs[TID=1080]: FFFFCE8A1038A010: >>Read 0000000000000000:000002231FC0B1A0, Address=0000022321800000, Offset=0:0, Length=524288, Key=0
    *** Read: Offset: 0 Length: 524288
    *** Read: fh: 0 nodeid: 3
    *** VirtFsFuseRequest: >>req: 15 unique: 29 len: 524328
    virtiofs[TID=1080]: FFFFCE8A1038A010: <<Read IoStatus=c0000001[0]
    virtiofs[TID=1080]: FFFFCE8A0810C4D0: >>Close 0000000000000000:000002231FC0B1A0
    *** Close: fh: 0 nodeid: 3
    *** VirtFsFuseRequest: >>req: 18 unique: 30 len: 64
    virtiofs[TID=04c8]: FFFFCE8A10E79780: >>Close 0000000000000000:000002231FC0B110
    *** Close: fh: 1 nodeid: 3
    *** VirtFsFuseRequest: >>req: 18 unique: 31 len: 64
    virtiofs[TID=0ac8]: FFFFCE8A112F5C10: >>Close 0000000000000000:000002231FC0AF30
    *** Close: fh: 0 nodeid: 4
    *** VirtFsFuseRequest: >>req: 29 unique: 32 len: 64
    *** VirtFsFuseRequest: <<len: 16 error: 0 unique: 30
    virtiofs[TID=1080]: FFFFCE8A0810C4D0: <<Close IoStatus=0[0]
    *** VirtFsFuseRequest: <<len: 16 error: 0 unique: 31
    virtiofs[TID=04c8]: FFFFCE8A10E79780: <<Close IoStatus=0[0]
    *** VirtFsFuseRequest: <<len: 16 error: 0 unique: 32
    virtiofs[TID=0ac8]: FFFFCE8A112F5C10: <<Close IoStatus=0[0]

</details>

Let me know if there's anything else I can do to help diagnose/debug/test.

Thanks,
Kevin

**Comment by hammerg** [comment]
Thanks for bug report. I was unable to reproduce it with both large ISO file and bigfile.bin (as you described). I also don't see any errors in the attached log file.

Which file system are you using on the host shared folder?

```
C:\Temp>dir z:\*.iso
 Volume in drive Z is myfs
 Volume Serial Number is 0000-0000

 Directory of z:\

14/05/2020  08:58     5,650,477,056 en_windows_10_consumer_editions_version_2004_x64_dvd_8d28c5d7.iso
16/07/2020  10:59     5,419,753,472 en_windows_server_2019_updated_june_2020_x64_dvd_7757177c.iso
               2 File(s) 11,070,230,528 bytes
               0 Dir(s)  44,777,963,520 bytes free

C:\Temp>copy z:\en_windows_10_consumer_editions_version_2004_x64_dvd_8d28c5d7.iso .
        1 file(s) copied.

C:\Temp>dir *.iso
 Volume in drive C has no label.
 Volume Serial Number is F8B4-BE6D

 Directory of C:\Temp

14/05/2020  08:58     5,650,477,056 en_windows_10_consumer_editions_version_2004_x64_dvd_8d28c5d7.iso
               1 File(s)  5,650,477,056 bytes
               0 Dir(s)  51,323,863,040 bytes free
````

Thanks,

    Gal.

**Comment by kevinoid** [comment]
> Which file system are you using on the host shared folder?

I tested with `tmpfs` and `ext4` on the host, the issue appeared on both.

Is there a standardized test process or CI that you use for virtio-fs?  If so, I could attempt to reproduce it there.  (If not, perhaps I could contribute to setting one up?)

**Comment by YanVugenfirer** [comment]
Hi Kevin,

We run MS certification tests as CI. 
For that, we use https://github.com/hck-ci that can orchestrate VMs, networking, storage, and services needed on the host for each device, and then it runs HCK\HLK tests using MS API.
We can create a "tester" definition that will run some other custom code on the guest, so virtio-fs specific CI tests will be welcomed.

Best regards,
Yan.

**Comment by YanVugenfirer** [comment]
+ @Basils

**Comment by kevinoid** [comment]
Thanks @YanVugenfirer, that looks great!  Unfortunately, I don't think I'll have time to work out how to run AutoHCK and define the "tester" you suggested today.  It seems non-trivial, to say the least.

Any ideas how to share a reproduction?  Would it be helpful if I sent a disk image (of a fresh, non-activated VM) with instructions to demonstrate the issue?

**Comment by YanVugenfirer** [comment]
@kevinoid Regarding CI - actually the "tester" part is something we can do. From you, we will need a script or an application that runs on guest and instructions for the host setup.

Regarding reproduction (@hammerg) - sure, the VM image will be definitely very helpful.

**Comment by kevinoid** [comment]
Sounds great.  I'll work on preparing of those.  Thanks again @YanVugenfirer!

**Comment by kevinoid** [comment]
I've come up with a full reproduction:
1. Download and decompress the [disk image](https://mega.nz/file/oQhCVRaJ#hPRzXwX8hJP2Gk6i7tClJRE2xBawQ6CbwbWJQMbTkkk).  (Or follow the [steps to create the disk image](https://gist.github.com/kevinoid/583dfee8355b638572eb907c5ae8e06c).)
2. Launch QEMU with `bigfile.bin` in a virtiofs share:
   ```sh
   #!/bin/sh
   set -Ceux
   # Create directory with test file to share
   share="$(mktemp -d)"
   dd if=/dev/zero of="$share/bigfile.bin" bs=2097153 count=1
   # Start virtiofsd (as root) with socket owned by invoking user
   virtiofs_sock="$(mktemp -ut virtiofs-XXXXXX).sock"
   sudo sh -Ceuc "
   nohup /usr/lib/qemu/virtiofsd --debug --socket-path='$virtiofs_sock' -o source='$share' >|virtiofsd.log 2>&1 &
   sleep 1	# Wait for virtiofsd to create $virtiofs_sock
   chown '$(id -u)' '$virtiofs_sock'"
   qemu_exit=0
   qemu-system-x86_64 \
   	-no-user-config \
   	-nodefaults \
   	-machine pc-q35-5.0,accel=kvm \
   	-cpu IvyBridge-IBRS \
   	-m 2G \
   	-object memory-backend-memfd,id=mem,size=2G,share=yes \
   	-numa node,memdev=mem \
   	-vga std \
   	-hda win10-virtiofs.qcow2 \
   	-chardev "socket,id=char0,path=$virtiofs_sock" \
   	-device vhost-user-fs-pci,chardev=char0,tag=myfs \
   	|| qemu_exit=$?
   # Clean up
   rm -rf "$share"
   exit "$qemu_exit"
   ```
3. In the guest, double-click `virtiofs.exe` on the desktop.
4. Run `copy Z:\bigfile.bin .` in `cmd.exe`, which should print:

       A device attached to the system is not functioning.
               0 file(s) copied.

While creating the reproduction, I noticed that the error **does not occur** if `queue-size=1024` is added to `-device ...myfs`.

If there's anything else I can do, let me know.  (Unless you object, I'll open a separate issue to discuss test script ideas, once I come up with some.)

**Comment by menli820** [comment]
@kevinoid Thanks for report this issue, I am the virtio-fs QE . @hammerg  I  also reproduce this issue on downstream, Do I need to file a bug  to track this issue?

reproduce steps:
1. start virtiofsd daemon on host
#/usr/libexec/virtiofsd --socket-path=/tmp/vhostqemu -o source=/tmp/virtiofs_test -o cache=always

2. boot up win1064 guest with virtiofs device.

-m 4G \
-object memory-backend-file,id=mem,size=4G,mem-path=/dev/shm,share=on \
-numa node,memdev=mem \
-chardev socket,id=char0,path=/tmp/vhostqemu \
-device vhost-user-fs-pci,chardev=char0,tag=myfs \ 

3. install winfsp tool and virtiofs driver in guest

4. run virtiofs.exe
virtiofs.exe -D c:\virtiofs-log.txt -d -1

3. on host side：
dd if=/dev/zero of=bigfile.bin bs=2048M count=1

4. on the guest:
Run copy Z:\bigfile.bin . in cmd.exe

Actual result:
after step4, file not copy successfully.

pkg:
qemu-kvm-5.1.0-17.module+el8.3.1+9213+7ace09c3.x86_64
seabios-bin-1.14.0-1.module+el8.3.0+7638+07cf13d2.noarch
kernel-4.18.0-240.el8.x86_64
virtio-win-prewhql-0.1-191.iso

additional info:
can not reproduce it with '-device vhost-user-fs-pci,queue-size=1024,chardev=char0,tag=myfs '

**Comment by menli820** [comment]
> @kevinoid Thanks for report this issue, I am the virtio-fs QE . @hammerg I also reproduce this issue on downstream, Do I need to file a bug to track this issue?
> 
> reproduce steps:
> 
> 1. start virtiofsd daemon on host
>    #/usr/libexec/virtiofsd --socket-path=/tmp/vhostqemu -o source=/tmp/virtiofs_test -o cache=always
> 2. boot up win1064 guest with virtiofs device.
> 
> -m 4G
> -object memory-backend-file,id=mem,size=4G,mem-path=/dev/shm,share=on
> -numa node,memdev=mem
> -chardev socket,id=char0,path=/tmp/vhostqemu
> -device vhost-user-fs-pci,chardev=char0,tag=myfs \
> 
> 1. install winfsp tool and virtiofs driver in guest
> 2. run virtiofs.exe
>    virtiofs.exe -D c:\virtiofs-log.txt -d -1
> 3. on host side：
>    dd if=/dev/zero of=bigfile.bin bs=2048M count=1
> 4. on the guest:
>    Run copy Z:\bigfile.bin . in cmd.exe
> 
> Actual result:
> after step4, file not copy successfully.
> 
> pkg:
> qemu-kvm-5.1.0-17.module+el8.3.1+9213+7ace09c3.x86_64
> seabios-bin-1.14.0-1.module+el8.3.0+7638+07cf13d2.noarch
> kernel-4.18.0-240.el8.x86_64
> virtio-win-prewhql-0.1-191.iso
> 
> additional info:
> can not reproduce it with '-device vhost-user-fs-pci,queue-size=1024,chardev=char0,tag=myfs '

track this bug with: https://bugzilla.redhat.com/show_bug.cgi?id=1873088

**Comment by chipzhang** [comment]
I can reproduce this issue as well.

Host:
Ubuntu 20.04, with kernel 5.4.0-52-generic, with qemu/libvirt installed from Ubuntu server backports PPA.
`qemu-system-x86_64 --version`:
`QEMU emulator version 5.2.0 (Debian 1:5.2+dfsg-9ubuntu3.1~backport20.04-202107160036~ubuntu20.04.1)`
`libvirtd --version`:
`libvirtd (libvirt) 7.0.0`

Guest:
Windows 10 LTSC 2019

XML:
```
...
<memoryBacking>
  <source type='memfd'/>
  <access mode='shared'/>
</memoryBacking>
...
<filesystem type="mount" accessmode="passthrough">
  <driver type="virtiofs"/>
  <source dir="/share/path"/>
  <target dir="share-name"/>
</filesystem>
...
```

**Comment by kinifwyne** [comment]
I had a similar problem, adding the following to the device XML solved the problem.

```xml
...
<filesystem type="mount" accessmode="passthrough">
  <driver type="virtiofs" queue="1024"/>
  ...
</filesystem>
...
```

**Comment by YanVugenfirer** [comment]
Hello All,

Please help us understanding you use cases for using virtio-fs, and thus make us virtio-fs support better.
Please participate in the discussion and add your use cases: https://github.com/virtio-win/kvm-guest-drivers-windows/discussions/726

Thanks a lot,
Yan.

**Comment by viktor-prutyanov** [comment]
It seems that `queue="1024"` helps.
